### PR TITLE
Add NMS iou threshold configurable parameter

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/common/configs/training_base.py [hunk 1]
```diff
@@ -231,6 +231,15 @@ class BasePostprocessing(ParameterGroup):
             affects_outcome_of=ModelLifecycle.INFERENCE,
         )
 
+        nms_iou_threshold = configurable_float(
+            default_value=0.0,
```
Comment (sovrasov) [inline src/otx/algorithms/common/configs/training_base.py]: Why do we have 0 here? Should it be ~0.5?

File: src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 img_size = (512, 512)
 
 tile_cfg = dict(
-    tile_size=400, min_area_ratio=0.9, overlap_ratio=0.2, iou_threshold=0.45, max_per_img=1500, filter_empty_gt=True
+    tile_size=400, min_area_ratio=0.9, overlap_ratio=0.2, iou_threshold=0.7, max_per_img=1500, filter_empty_gt=True
```
Comment (sovrasov) [inline src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py]: How do we know that 0.7 is better?

**Comment by sovrasov** [comment]
Are detectors so sensitive to IoU in NMS? We shouldn't apply any default value tricks, since it all is visible in the UI.
My initial thought was to bind 0.5 default value for all detectors and tilers, conduct a couple experiments, and call it a day. Unfortunately, that's a sacrifice we're making because of the imperfect Geti configuration system, which doesn't support per model tuning for default params.

If that's not possible (the detectors are too sensitive, we could think of changing the tiling threshold only)

File: src/otx/algorithms/common/configs/training_base.py (hunk 1 above)
Comment (jaegukhyun) [reply src/otx/algorithms/common/configs/training_base.py]: Every detectors have their own nms iou threshold in their model.py. It is different across the architecture, So here default value 0.0, which means we're going to use the default value of object detector. 
In addition, this methods can prevent confusion from OTX side. Currently OTX do not know whether hyperparameters come from defaults or users command. So this way can easily determine whether the nms iou threshold come from users or defaults confiuration.

File: src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py (hunk 2 above)
Comment (jaegukhyun) [reply src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py]: This is for sync with models' parameter. I thought our plan is 
1. Sync model and tiler's nms iou threshold
2. Conduct exp for verifying it
@sovrasov @eugene123tw Feel free to correct me.

File: src/otx/algorithms/common/configs/training_base.py (hunk 1 above)
Comment (sovrasov) [reply src/otx/algorithms/common/configs/training_base.py]: The problem that in Geti UI, users would see 0.0 as IoU threshold, which looks like incorrect value.

File: src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py (hunk 2 above)
Comment (sovrasov) [reply src/otx/algorithms/detection/configs/base/data/tiling/efficientnet_iseg_tile_pipeline.py]: That's right, the only problem is the default value.

**Comment by eugene123tw** [comment]
> Are detectors so sensitive to IoU in NMS? We shouldn't apply any default value tricks, since it all is visible in the UI. My initial thought was to bind 0.5 default value for all detectors and tilers, conduct a couple experiments, and call it a day. Unfortunately, that's a sacrifice we're making because of the imperfect Geti configuration system, which doesn't support per model tuning for default params.
> 
> If that's not possible (the detectors are too sensitive, we could think of changing the tiling threshold only)

@jaegukhyun Based on the experiment: https://intel-my.sharepoint.com/:x:/p/eugene_liu/EWr2Xk17brBDgqZO62qlgSIBNNugJIalJZTWQJYqmaNiVg?e=SgQqpe

Let's change the default NMS IoU threshold to 0.5, what do you think?

File: src/otx/algorithms/detection/adapters/mmdet/configurer.py [hunk 3]
```diff
@@ -61,6 +62,30 @@ def configure_max_num_detections(self, cfg, max_num_detections):
                 train_cfg.rpn_proposal.nms_pre = max_num_detections * 20
                 train_cfg.rpn_proposal.max_per_img = max_num_detections * 10
 
+    def configure_nms_iou_threshold(self, cfg, nms_iou_threshold):
+        """Configure nms iou threshold to user specified value if the object detector uses nms."""
+        if nms_iou_threshold > 0:
```
Comment (eugene123tw) [inline src/otx/algorithms/detection/adapters/mmdet/configurer.py]: Do we need this condition if it's always > 0?

File: src/otx/algorithms/detection/configs/detection/configuration.yaml [hunk 4]
```diff
@@ -294,6 +294,26 @@ postprocessing:
...
+    min_value: 0
+    type: FLOAT
+    ui_rules:
+      action: DISABLE_EDITING
+      operator: AND
+      rules: []
+      type: UI_RULES
+    value: 0.01
+    visible_in_ui: true
+    warning: null
```
Comment (eugene123tw) [inline src/otx/algorithms/detection/configs/detection/configuration.yaml]: Where do we add the warning for retraining?

**Review by eugene123tw — COMMENTED** [review]
Only some minor comments

File: src/otx/algorithms/detection/configs/detection/configuration.yaml (hunk 4 above)
Comment (jaegukhyun) [reply src/otx/algorithms/detection/configs/detection/configuration.yaml]: I updated that info in description

Comment (jaegukhyun) [reply src/otx/algorithms/detection/configs/detection/configuration.yaml]: I moved it to warning

File: src/otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 3 above)
Comment (jaegukhyun) [reply src/otx/algorithms/detection/adapters/mmdet/configurer.py]: Removed!

**Review by eugene123tw — DISMISSED** [review]
Thanks Jaeguk!

**Review by sovrasov — DISMISSED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by eugene123tw — APPROVED** [review]

**Pull request merged**